### PR TITLE
firefox: wrapper: add libva to libs

### PR DESCRIPTION
**Pull request by colemickens**
#### Motivation for this change

This change allows VA-API to work in Firefox Nightly builds as provided by `nixpkgs-mozilla`.

Two changes:
* add `libva` to the Firefox wrapper
* move to `ffmpeg_4` (without it, `(firefox/ffmpeg)` doesn't attempt to use libva/vaapi)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by colemickens** [comment]
Followup:

* upstream's testing instructions: https://bugzilla.mozilla.org/show_bug.cgi?id=1610199#c31
* I suffer from the symptoms in this bug, however: https://bugzilla.mozilla.org/show_bug.cgi?id=1619543

**Comment by colemickens** [comment]
Looks like it's probably related to this:

```
[Child 23304: MediaPDecoder #1]: D/PlatformDecoderModule Initialising FFmpeg decoder.
[Child 23304: MediaPlayback #1]: D/PlatformDecoderModule Sandbox decoder rejects requested type
[Child 23304: MediaPDecoder #3]: D/PlatformDecoderModule Initialising VA-API FFmpeg decoder
[Child 23304: MediaPDecoder #3]: D/PlatformDecoderModule VA-API FFmpeg is disabled by platform
[Child 23304: MediaPDecoder #3]: D/PlatformDecoderModule Initialising FFmpeg decoder.
[Child 23304: MediaPlayback #1]: D/PlatformDecoderModule Sandbox decoder rejects requested type
[Child 23304: MediaPlayback #1]: D/PlatformDecoderModule Sandbox decoder rejects requested type
[Child 23304: MediaPlayback #1]: D/PlatformDecoderModule Sandbox decoder rejects requested type
[Child 23304: MediaPlayback #1]: D/PlatformDecoderModule Sandbox decoder rejects requested type
[Child 23304: MediaPDecoder #2]: D/PlatformDecoderModule Initialising FFmpeg decoder.
[Child 23304: MediaPDecoder #4]: D/PlatformDecoderModule Initialising VA-API FFmpeg decoder
[Child 23304: MediaPDecoder #4]: D/PlatformDecoderModule VA-API FFmpeg is disabled by platform
[Child 23304: MediaPDecoder #4]: D/PlatformDecoderModule Initialising FFmpeg decoder.
[Child 23304: MediaPDecoder #1]: D/PlatformDecoderModule FFmpeg init successful.
```

```
WebGL creation failed: 
* Refused to create WebGL2 context because of blacklist entry: FEATURE_FAILURE_GLXTEST_FAILED

GLXtest process failed (exited with status 1): No visuals found
```

**Comment by colemickens** [comment]
cc: @worldofpeace You replied in another thread about Firefox and hardware acceleration in NixOS. I am using the nightly package (a bin package).

**Comment by chpio** [comment]
> ```
> 23304: MediaPDecoder #3]: D/PlatformDecoderModule VA-API FFmpeg is disabled by platform
> ```

Is this maybe related? https://bugzilla.mozilla.org/show_bug.cgi?id=1610199#c35

**Comment by primeos** [comment]
@colemickens since Firefox 75 is released now and 76 should already be in the beta channel you might want to give this another shot, if you have time. If not feel free to simply ignore this comment ;)

It would also be also nice if we had VA-API support for the `firefox` package that we build from source, but that seems outside the scope of this PR (and Firefox rebuilds most likely take a very long time).

**Comment by colemickens** [comment]
Oops. In one version of this I had actually built firefox as well. Anyway, this works with Nightly!

```
[Child 29119: MediaPlayback #1]: D/PlatformDecoderModule Sandbox decoder rejects requested type
[Child 29119: MediaPlayback #1]: D/PlatformDecoderModule Sandbox decoder rejects requested type
[Child 29119: MediaPDecoder #1]: D/PlatformDecoderModule Initialising VA-API FFmpeg decoder
[Child 29119: MediaPDecoder #2]: D/PlatformDecoderModule Initialising FFmpeg decoder.
libva info: VA-API version 1.6.0
libva info: Trying to open /run/opengl-driver/lib/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_6
libva info: va_openDriver() returns 0

[h264 @ 0x7f4fe222e000] nal_unit_type: 7(SPS), nal_ref_idc: 3
[h264 @ 0x7f4fe222e000] nal_unit_type: 8(PPS), nal_ref_idc: 3
[Child 29119: MediaPDecoder #2]: D/PlatformDecoderModule VA-API FFmpeg init successful
[h264 @ 0x7f4fe222e000] nal_unit_type: 6(SEI), nal_ref_idc: 0
[h264 @ 0x7f4fe222e000] nal_unit_type: 5(IDR), nal_ref_idc: 3
[Child 29119: MediaPDecoder #2]: D/PlatformDecoderModule Choosing FFmpeg pixel format for VA-API video decoding.
[Child 29119: MediaPDecoder #2]: D/PlatformDecoderModule Requesting pixel format VAAPI_VLD
[h264 @ 0x7f4fe222e000] Format vaapi_vld chosen by get_format().
[h264 @ 0x7f4fe222e000] Format vaapi_vld requires hwaccel initialisation.
[h264 @ 0x7f4fe222e000] Considering format 0x3231564e -> nv12.
[h264 @ 0x7f4fe222e000] Picked nv12 (0x3231564e) as best match for yuv420p.
```

We can merge? In the meantime I can try to make the change for `firefox` from source too.

**Comment by colemickens** [comment]
Ah, I should be careful, I actually have some other related firefox commits on the branch that I tested from. Let me do a few additional checks around nightly and/or wait for this build to finish.

**Comment by colemickens** [comment]
Interesting, it looks like this does require moving over to ffmpeg4!

**Comment by colemickens** [comment]
This time I'm sure I'm testing firefox-nightly with the wrapper from this branch:
```fish
export MOZ_LOG="PlatformDecoderModule:5"; \
 set f (~/code/nixcfg/nixbuild.sh -A latest.firefox-nightly-bin); \
$f/bin/firefox --no-remote -P vaapi &> /tmp/firefox
```

similar logs as I posted ~1 hr ago, with two commits now, pushing now.

**Comment by colemickens** [comment]
Note, I don't know if this is a strong requirement by Firefox or is due to how we build ffmpeg4? But I definitely couldn't get this working with nightly without adding the ffmpeg4 commit.

File: pkgs/applications/networking/browsers/firefox/common.nix
Comment (andir) [inline pkgs/applications/networking/browsers/firefox/common.nix:26]: This should probably be version gated. The old esr version doesn't support this I guess?

We would default that to `true` with the next ESR release.

File: pkgs/applications/networking/browsers/firefox/wrapper.nix
Comment (andir) [inline pkgs/applications/networking/browsers/firefox/wrapper.nix]: Shouldn't that be the same ffmpeg version we are using in the build?

File: pkgs/applications/networking/browsers/firefox/wrapper.nix
Comment (andir) [inline pkgs/applications/networking/browsers/firefox/wrapper.nix:68]: Is this stil required given that we are already building firefox with that lib?

File: pkgs/applications/networking/browsers/firefox/common.nix
Comment (primeos) [reply pkgs/applications/networking/browsers/firefox/common.nix:26]: Might make sense (e.g. to make it clear which versions support this), but since this only affect `LD_LIBRARY_PATH` it shouldn't be a problem for older versions AFAIK (only provides an additional library that can simply be ignored).

File: pkgs/applications/networking/browsers/firefox/wrapper.nix
Comment (primeos) [reply pkgs/applications/networking/browsers/firefox/wrapper.nix]: I was already wondering about this, because only the wrapper depends on `ffmpeg`. Do we build Firefox with a bundled version from Mozilla? Or did I miss something?

Apart from that `ffmpeg_4` would be nice to have in general :) E.g. Arch also builds with 4.x

File: pkgs/applications/networking/browsers/firefox/wrapper.nix
Comment (primeos) [reply pkgs/applications/networking/browsers/firefox/wrapper.nix:68]: On `master` I don't see a direct dependency on `libva`.
(@colemickens mentioned that this is currently for the binary package only and I don't think our source build supports VA-API yet)

File: pkgs/applications/networking/browsers/firefox/common.nix
Comment (andir) [reply pkgs/applications/networking/browsers/firefox/common.nix:26]: This isn't the wrapper. This is in the build expression for firefox.

File: pkgs/applications/networking/browsers/firefox/wrapper.nix
Comment (andir) [reply pkgs/applications/networking/browsers/firefox/wrapper.nix:68]: This PR adds that support: https://github.com/NixOS/nixpkgs/pull/81917/files#diff-b39b9a678f9ecf5d0e1581832f7bf99dR26

Comment (primeos) [reply pkgs/applications/networking/browsers/firefox/wrapper.nix:68]: Oh, sorry. My bad, I didn't catch up with the third commit (also in the first change).

Comment (colemickens) [reply pkgs/applications/networking/browsers/firefox/wrapper.nix:68]: If I drop it from the wrapper it seems like it can't find it at runtime?
`   1 [Child 20523: MediaPDecoder #2]: D/PlatformDecoderModule libva library is missing`

File: pkgs/applications/networking/browsers/firefox/wrapper.nix
Comment (colemickens) [reply pkgs/applications/networking/browsers/firefox/wrapper.nix]: I asked in #build:mozilla.org. Apparently they don't require an ffmpeg to build (which is true, based on our derivation). My guess is the thing in tree is a wrapper, but I didn't look; I'm afraid of falling down a rabbit hole.

**Comment by colemickens** [comment]
Alright, I'm going to put a couple more hours in and then give up and back out the change to the source package and leave this as a change only targetting nightly. However, I know a lot of people want the source build to have va-api too.

**Comment by colemickens** [comment]
Okay, back to the original scope of this PR. I'll iterate on this, but I know this helps Nightly users, so let's go with this.

I can confirm both commits are needed. The older ffmpeg in the wrapper does NOT work, it doesn't even attempt to use va-api or libva.

Fortunately, this also means it's easy for Nightly users to test, since this only requires a change to the wrapper and won't need to rebuild Firefox.

File: pkgs/applications/networking/browsers/firefox/common.nix
Comment (colemickens) [reply pkgs/applications/networking/browsers/firefox/common.nix]: Sorry about the confusion here. I'd never meant to push the source build changes publicly and was hoping I could just get it working. I've decided instead to just keep the changes to the wrapper for this PR. Going to resolve this accordingly.

**Review by primeos — APPROVED** [review]
LGTM :) But I'll leave the final decision and merging up to @andir since he's the maintainer.

**Comment by andir** [comment]
Can I test this with the firefox-{beta,bin} from nixpkgs or will that not work?

**Pull request merged**
